src/utils.h

```diff
@@ -259,6 +265,65 @@ ERR r_try_catch(void (*fn)(void*),
                 void (*hnd)(void*),
                 void* hnd_data);
 
+extern SEXP vctrs_ns_env;
+extern SEXP syms_cnd_signal;
+static inline void r_cnd_signal(SEXP cnd) {
+  SEXP call = PROTECT(Rf_lang2(syms_cnd_signal, cnd));
+  Rf_eval(call, vctrs_ns_env);
+  UNPROTECT(1);
+}
+
+extern SEXP result_attrib;
+
+// Inputs are protected so this can be used as return value
+static inline SEXP r_result(SEXP x, ERR err) {
+  PROTECT(x);
+  if (!err) {
+    err = R_NilValue;
+  }
+  PROTECT(err);
+
+  SEXP result = PROTECT(Rf_allocVector(VECSXP, 2));
+  SET_VECTOR_ELT(result, 0, x);
+  SET_VECTOR_ELT(result, 1, err);
+
+  SET_ATTRIB(result, result_attrib);
+  SET_OBJECT(result, 1);
+
+  UNPROTECT(3);
+  return result;
+}
+
+// Inputs are protected so this can be used as return value
```

Comment:
Does `x` explicitly need to be `PROTECT()`ed? I wasn't sure how this works

Reply:
Not sure I'm happy with the non-standard semantics, but it seemed to make sense since these functions are always tail calls. I thought we'd make an experiment of it and switch to normal protection later if we don't like it.

Reply:
I've changed my mind about this, these "result" operations are not common enough to be worth it. And should never be, since result objects are mostly temporary for prototyping C code in R.